### PR TITLE
Added duplicate tab feature

### LINKED ISSUE #524: Duplicate tab feature
Should preserve current left pane state, search box input, web view scroll position.

Firefox uses <kbd>Ctrl+Alt+T</kbd> as a shortcut.

### PR DESCRIPTION
**Pull request by innermous**
Closes #524 
Maybe any remarks?

### CODE REVIEW

**Comment by trollixx** [comment]
Thanks for the contribution!

I can see a few minor issues, that should be pretty easy to fix. I'll add detailed comments in the code. Also, I am planning to push a few changes I have, so this PR might need a rebase.

File: src/ui/mainwindow.cpp [hunk 1]
```diff
@@ -102,6 +114,10 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
     connect(focusSearch, &QShortcut::activated,
             ui->lineEdit, static_cast<void (SearchEdit::*)()>(&SearchEdit::setFocus));
 
+    QShortcut *duplicate = new QShortcut(QKeySequence(QStringLiteral("Ctrl+Alt+T")), this);
+    duplicate->setContext(Qt::ApplicationShortcut);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: No need in `Qt::ApplicationShortcut`.

File: src/ui/mainwindow.cpp [hunk 2]
```diff
@@ -102,6 +114,10 @@ MainWindow::MainWindow(Core::Application *app, QWidget *parent) :
     connect(focusSearch, &QShortcut::activated,
             ui->lineEdit, static_cast<void (SearchEdit::*)()>(&SearchEdit::setFocus));
 
+    QShortcut *duplicate = new QShortcut(QKeySequence(QStringLiteral("Ctrl+Alt+T")), this);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: `QKeySequence` can be constructed implicitly.

File: src/ui/mainwindow.cpp [hunk 3]
```diff
@@ -455,6 +471,36 @@ void MainWindow::createTab()
     m_tabBar->setCurrentIndex(index);
 }
 
+void MainWindow::duplicateTab()
+{
+    saveTabState();
+    TabState* newTab = new TabState(*m_currentTabState);
+    const QString& parentTitle = m_tabBar->tabText(m_tabBar->currentIndex());
+    const QUrl& parentUrl = m_currentTabState->webPage->mainFrame()->url();
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: Cannot be compiled with Qt WebEngine.

File: src/ui/mainwindow.cpp [hunk 4]
```diff
@@ -455,6 +471,36 @@ void MainWindow::createTab()
...
+#else
+    newTab->webPage->setLinkDelegationPolicy(QWebPage::DelegateExternalLinks);
+    newTab->webPage->setNetworkAccessManager(m_application->networkManager());
+    newTab->webPage->mainFrame()->load(parentUrl);
+#endif
+
+    ui->treeView->verticalScrollBar()->setValue(m_currentTabState->searchScrollPosition);
+    ui->tocListView->verticalScrollBar()->setValue(m_currentTabState->tocScrollPosition);
+
+    m_tabStates.append(newTab);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: This breaks if the tab to duplicate is not the last one. Indices of states in `m_tabStates` have to match their tabs.

File: src/ui/mainwindow.cpp [hunk 5]
```diff
@@ -455,6 +471,36 @@ void MainWindow::createTab()
     m_tabBar->setCurrentIndex(index);
 }
 
+void MainWindow::duplicateTab()
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: A lot of code is taken from `MainWindow::createTab()`. It would be nice to avoid this code duplication.

Comment (innermous) [inline src/ui/mainwindow.cpp]: Aint it gonna get messy a bit if we combine two functionalities in one method?

File: src/ui/mainwindow.cpp (hunk 4 above)
Comment (innermous) [inline src/ui/mainwindow.cpp]: Can you explain, what you mean? It works correctly as it should, duplicated tabs indices match their position in tab bar...

File: src/ui/mainwindow.cpp [hunk 6]
```diff
@@ -455,6 +458,24 @@ void MainWindow::createTab()
...
+{
+    saveTabState();
+    TabState* previous = currentTabState();
+
+    createTab();
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: I don't like this approach, creating something just to override. I am thinking about extending `createTab()` with the necessary functionality...

**Comment by trollixx** [comment]
An important piece appears to be missing - browser history. The new tab should preserve history of the originating tab.

**Comment by innermous** [comment]
Added history preserving. Still have no idea how to implement all in 1 method though.
Only possible outcome that I see - extend createTab with bool duplicate. But that yields lots of if statements as I see it.
If you have any thought - share please.

File: src/ui/mainwindow.cpp [hunk 7]
```diff
@@ -474,6 +501,28 @@ void MainWindow::createTab()
...
+    *newTab = *previous;
+
+#ifdef USE_WEBENGINE
+    newTab->webPage->setUrl(previous->webPage->url());
+#else
+    newTab->webPage->mainFrame()->setUrl(previous->webPage->mainFrame()->url());
+#endif
+
+    QByteArray historyArray;
+    previous->saveHistory(historyArray);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: I think `TabState`'s copy constructor/assignment should take care of history.

**Comment by trollixx** [comment]
@innermous any reason not to rebase this on top of master? I've made some changes to minify the tab related code. `duplicateTab(int index)` where `index` belongs to the tab to duplicate, would work great. Now, with the minimised amount of code there's no need to merge `createTab()` and `duplicateTab()`.

**Comment by innermous** [comment]
@trollixx sorry, i was trying to understand how to make this work after you removed reloadTabState, but I don't know really.

**Comment by trollixx** [comment]
It should just work automatically, after you call `m_tabBar->setCurrentIndex(index)`.

**Comment by innermous** [comment]
@trollixx i don't understand what's the point of index in this case, if we duplicate current tab.

File: src/ui/forms/mainwindow.ui [hunk 8]
```diff
@@ -140,7 +140,8 @@
               <string>←</string>
              </property>
              <property name="icon">
-              <iconset theme="go-previous"/>
+              <iconset theme="go-previous">
+               <normaloff>.</normaloff>.</iconset>
```
Comment (trollixx) [inline src/ui/forms/mainwindow.ui]: This and similar changes below are irrelevant.

File: src/ui/forms/mainwindow.ui [hunk 9]
```diff
@@ -311,12 +318,18 @@
...
    <property name="icon">
-    <iconset theme="edit-find"/>
+    <iconset theme="edit-find">
+     <normaloff>.</normaloff>.</iconset>
    </property>
    <property name="text">
     <string>&amp;Find</string>
    </property>
   </action>
+  <action name="actionDuplicateTab">
```
Comment (trollixx) [inline src/ui/forms/mainwindow.ui]: I've checked with other browsers, and neither has `Duplicate Tab` in the main menu. I think there's no need to add it here. Its place is tab's context menu, which is yet to be implemented. Just a shortcut should be fine for now.

File: src/ui/mainwindow.cpp [hunk 10]
```diff
@@ -468,6 +514,22 @@ void MainWindow::createTab(int index)
     m_tabBar->setCurrentIndex(index);
 }
 
+void MainWindow::duplicateTab(int index)
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: Let's make `index` a required parameter here, and actually duplicate the specific tab. That will allow duplicating tabs from a context menu, once we have it.

File: src/ui/mainwindow.cpp [hunk 11]
```diff
@@ -468,6 +514,22 @@ void MainWindow::createTab(int index)
...
+    if (index == -1)
+        index = m_tabStates.size();
+
+    TabState* previous = currentTabState();
+    TabState* newTab = new TabState(*previous);
+
+    connect(newTab->searchModel, &SearchModel::queryCompleted, this, &MainWindow::queryCompleted);
+    connect(newTab->tocModel, &SearchModel::queryCompleted, this, &MainWindow::toggleToc);
+
+    m_tabStates.insert(index, newTab);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: The new tab should be placed after the one we duplicate.

File: src/ui/mainwindow.cpp [hunk 12]
```diff
@@ -85,6 +85,40 @@ struct TabState
...
+        restoreHistory(historyArray);
+    }
+
+    void restoreHistory(const QByteArray &data) const
+    {
+        QDataStream stream(data);
+        stream >> *webPage->history();
+    }
+
+    void saveHistory(QByteArray &data) const
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: Just for the sake of consistency, let's follow Qt's pattern (see `QSplitter::saveState()`, `QByteArray QWidget::saveGeometry()`) here:

```cpp
    QByteArray saveHistory() const
    {
        QByteArray history;
        QDataStream stream(&history, QIODevice::WriteOnly);
        stream << *webPage->history();
        return history;
    }
```

File: src/ui/mainwindow.cpp [hunk 13]
```diff
@@ -85,6 +85,40 @@ struct TabState
...
+        webPage->setLinkDelegationPolicy(QWebPage::DelegateExternalLinks);
+        webPage->setNetworkAccessManager(Core::Application::instance()->networkManager());
+#endif
+
+        QByteArray historyArray;
+        rhs.saveHistory(historyArray);
+        restoreHistory(historyArray);
+    }
+
+    void restoreHistory(const QByteArray &data) const
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]:
```cpp
void restoreHistory(const QByteArray &history) const
```
Nitpicking :)

**Comment by trollixx** [comment]
Sorry, it took me so long to get back. Things should go better from now on. Thank you for bearing with me! :)

File: src/ui/mainwindow.cpp (hunk 13 above)
Comment (innermous) [inline src/ui/mainwindow.cpp]: What's wrong in 
```cpp
 void restoreHistory(const QByteArray &history) const
```
?

File: src/ui/mainwindow.h [hunk 14]
```diff
@@ -93,6 +93,7 @@ private slots:
     void openDocset(const QModelIndex &index);
     void queryCompleted();
     void closeTab(int index = -1);
+    void duplicateTab(int index = -1);
```
Comment (trollixx) [inline src/ui/mainwindow.h]: There should be no default value now.

File: src/ui/mainwindow.cpp [hunk 15]
```diff
@@ -468,6 +515,19 @@ void MainWindow::createTab(int index)
     m_tabBar->setCurrentIndex(index);
 }
 
+void MainWindow::duplicateTab(int index)
+{
+    TabState* previous = m_tabStates.at(index++);
```
Comment (trollixx) [inline src/ui/mainwindow.cpp]: I'd still check whether the index is valid.

Comment (trollixx) [inline src/ui/mainwindow.cpp]: And please use `TabState *previous` style.

**Comment by trollixx** [comment]
A few tiny things above, a rebase, and it's good to go!

And please amend the commit message to `ui: Implement tab duplication (fixes #524)`. Thanks!

**Pull request merged**